client/src/main/java/io/pravega/client/netty/impl/SessionHandler.java

```diff
@@ -0,0 +1,241 @@
+        Preconditions.checkState(!disableSession.getAndSet(true), "Sessions are disabled, incorrect usage pattern.");
+        log.info("Creating a new connection with session disabled for Endpoint {}. The current Channel is {}.", connectionName, channel.get());
+        sessionIdReplyProcessorMap.put(SESSION_DISABLED, rp);
+        return new ClientConnectionImpl(connectionName, SESSION_DISABLED, batchSizeTracker, this);
+    }
+
+    /**
+     * Close a session. This is invoked when the ClientConnection is closed.
+     * @param clientConnection Client Connection.
+     */
+    public void closeSession(ClientConnection clientConnection) {
+        final ClientConnectionImpl clientConnectionImpl = (ClientConnectionImpl) clientConnection;
+        int session = clientConnectionImpl.getSession();
+        log.info("Closing Session: {} for Endpoint: {}", session, clientConnectionImpl.getConnectionName());
+        sessionIdReplyProcessorMap.remove(session);
+    }
+
+    /**
+     * Check the current status of Connection.
+     * @return True if the connection is established.
+     */
+    public boolean isConnectionEstablished() {
+        return channel.get() != null;
+    }
+
+    /**
+     * Fetch the netty channel. If {@link Channel} is null then throw a ConnectionFailedException.
+     * @return  The current {@link Channel}
+     * @throws ConnectionFailedException Throw if connection is not established.
+     */
+    Channel getChannel() throws ConnectionFailedException {
+        Channel ch = channel.get();
+        if (ch == null) {
+            throw new ConnectionFailedException("Connection to " + connectionName + " is not established.");
+        }
+        return ch;
+    }
+
+    /**
+     * Set the Recent Message flag. This is used to avoid sending redundant KeepAlives over the connection.
+     */
+    void setRecentMessage() {
+        recentMessage.set(true);
+    }
+
+    /**
+     * This function completes the input future when the channel is registered.
+     *
+     * @param future CompletableFuture which will be completed once the channel is registered.
+     */
+    void completeWhenRegistered(final CompletableFuture<Void> future) {
+        Preconditions.checkNotNull(future, "future");
+        registeredFutureLatch.register(future);
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        super.channelRegistered(ctx);
+        Channel ch = ctx.channel();
+        channel.set(ch);
+        log.info("Connection established with Endpoint: {} on ChannelId: {}.", connectionName, ch);
+        ch.write(new WireCommands.Hello(WireCommands.WIRE_VERSION, WireCommands.OLDEST_COMPATIBLE_VERSION), ch.voidPromise());
+        registeredFutureLatch.release(null); //release all futures waiting for channel registration to complete.
+        // WireCommands.KeepAlive messages are sent for every network connection to a SegmentStore.
```

Comment:
Do we need to send keepalives on a per session basis or per connection basis? I'd think the latter.

Reply:
Yes, KeepAlives are sent on a per connection basis.